### PR TITLE
Add IntelRealSenseCamera

### PR DESCRIPTION
**Pull request by Cadene**
## What this does

- Add intelrealsense.py in camera devices
- Use IntelRealSenseCamera as default for aloha.yaml robot config
- Can return depth map (with read and async_read)

TODO in https://github.com/huggingface/lerobot/pull/398:
- [ ] Add unit tests

## How it was tested

- Found the camera indices using intelrealsense.py and looking at images.
- Used the camera indices in aloha.yaml
- Ran data recording.
- Checked the image from cameras are well displayed during data recording.
- Checked the bgr color mode works.
- Checked `read` and `async_read` returns proper color_image and depth_map.
- Checked `del camera` works

```python
camera.read()
(array([[[0, 0, 2],
        [0, 0, 2],
        [1, 1, 3],
        ...,
        [0, 1, ..., 2, 1],
        [0, 2, 1]]], dtype=uint8), array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ...,...   [0, 0, 0, ..., 0, 0, 0]], dtype=uint16))
camera.async_read()
(array([[[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        ...,
        [0, 0, ..., 2, 1],
        [0, 2, 1]]], dtype=uint8), array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ...,...   [0, 0, 0, ..., 0, 0, 0]], dtype=uint16))
```

## How to checkout & try? (for the reviewer)

```bash
python lerobot/common/robot_devices/cameras/intelrealsense.py
```

```bash
python lerobot/scripts/control_robot.py record \
    --robot-path lerobot/configs/robot/aloha.yaml \
    --fps 30 \
    --root /tmp/data \
    --repo-id $USER/test \
    --num-episodes 1 \
    --run-compute-stats 0 \
    --warmup-time-s 2 \
    --episode-time-s 30 \
    --reset-time-s 1 \
    --force-override 1 \
    --push-to-hub 0

python lerobot/scripts/visualize_dataset_html.py \
  --root /tmp/data \
  --repo-id $USER/test
  ```

### CODE REVIEW

File: lerobot/common/robot_devices/cameras/intelrealsense.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
+"""
+This file contains ultilities for recording frames from Intel Realsense cameras.
```
Comment (aliberts) [inline lerobot/common/robot_devices/cameras/intelrealsense.py:2]:
```suggestion
This file contains utilities for recording frames from Intel Realsense cameras.
```

File: lerobot/common/robot_devices/cameras/intelrealsense.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+    if camera_ids is None:
+        camera_ids = find_camera_indices()
+
+    print("Connecting cameras")
+    cameras = []
+    for cam_idx in camera_ids:
+        camera = IntelRealSenseCamera(cam_idx, fps=fps, width=width, height=height)
+        camera.connect()
+        print(
+            f"IntelRealSense Camera({camera.camera_idx}, fps={camera.fps}, width={camera.width}, height={camera.height}, color_mode={camera.color_mode})"
```
Comment (shantanuparab-tr) [inline lerobot/common/robot_devices/cameras/intelrealsense.py:81]: 'IntelRealSenseCamera' object has no attribute 'camera_idx'. Did you mean: 'camera_index'
```suggestion
            f"IntelRealSense Camera({camera.camera_index}, fps={camera.fps}, width={camera.width}, height={camera.height}, color_mode={camera.color_mode})"
```

File: lerobot/common/robot_devices/cameras/intelrealsense.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+import threading
+import time
+import traceback
+from dataclasses import dataclass, replace
+from pathlib import Path
+from threading import Thread
+
+import cv2
+import numpy as np
+import pyrealsense2 as rs
```
Comment (aliberts) [inline lerobot/common/robot_devices/cameras/intelrealsense.py:18]: We should make sure that this is not executed when pyrealsense2 isn't installed (especially on MacOS)

File: lerobot/common/robot_devices/cameras/intelrealsense.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+    print(f"Saving images to {images_dir}")
+    frame_index = 0
+    start_time = time.perf_counter()
+    try:
+        with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
+            while True:
+                now = time.perf_counter()
+
+                for camera in cameras:
+                    # If we use async_read when fps is None, the loop will go full speed, and we will endup
```
Comment (aliberts) [inline lerobot/common/robot_devices/cameras/intelrealsense.py:101]:
```suggestion
                    # If we use async_read when fps is None, the loop will go full speed, and we will end up
```

File: lerobot/common/robot_devices/cameras/intelrealsense.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+    IntelRealSenseCameraConfig(30, 1280, 720)
+    IntelRealSenseCameraConfig(30, 640, 480, use_depth=True)
+    ```
+    """
+
+    fps: int | None = None
+    width: int | None = None
+    height: int | None = None
+    color_mode: str = "rgb"
+    use_depth: bool = False
```
Comment (aliberts) [inline lerobot/common/robot_devices/cameras/intelrealsense.py:150]: What shape (width, height, channel) does depth have?

File: lerobot/common/robot_devices/cameras/intelrealsense.py [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+
+class IntelRealSenseCamera:
+    """
+    The IntelRealSenseCamera class is similar to OpenCVCamera class but adds additional features for Intel Real Sense cameras:
+    - camera_index corresponds to the serial number of the camera,
+    - camera_index won't randomly change as it can be the case of OpenCVCamera for Linux,
+    - read is more reliable than OpenCVCamera,
+    - depth map can be returned.
+
+    To find the camera indices of your cameras, you can run our utility script that will be save a few frames for each camera:
```
Comment (aliberts) [inline lerobot/common/robot_devices/cameras/intelrealsense.py:174]:
```suggestion
    To find the camera indices of your cameras, you can run our utility script that will save a few frames for each camera:
```

**Review by aliberts — APPROVED** [review]
LGTM, thanks!

File: lerobot/common/robot_devices/cameras/intelrealsense.py [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+            is_camera_open = False
+            traceback.print_exc()
+
+        # If the camera doesn't work, display the camera indices corresponding to
+        # valid cameras.
+        if not is_camera_open:
+            # Verify that the provided `camera_index` is valid before printing the traceback
+            available_cam_ids = find_camera_indices()
+            if self.camera_index not in available_cam_ids:
+                raise ValueError(
```
Comment (shantanuparab-tr) [inline lerobot/common/robot_devices/cameras/intelrealsense.py:273]:
```suggestion
            if int(self.camera_index) not in available_cam_ids:
```

File: lerobot/common/robot_devices/cameras/intelrealsense.py (hunk 3 above)
Comment (Cadene) [reply lerobot/common/robot_devices/cameras/intelrealsense.py:18]: It won't be.

File: lerobot/common/robot_devices/cameras/intelrealsense.py [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+                    image = camera.read() if fps is None else camera.async_read()
+                    if image is None:
+                        print("No Frame")
+                    bgr_converted_image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
+
+                    executor.submit(
+                        save_image,
+                        bgr_converted_image,
+                        camera.camera_idx,
+                        frame_index,
```
[the remark this answers is not in the document]
Comment (Cadene) [reply lerobot/common/robot_devices/cameras/intelrealsense.py:112]: Wrong line no?

File: lerobot/common/robot_devices/cameras/intelrealsense.py (hunk 7 above)
Comment (Cadene) [reply lerobot/common/robot_devices/cameras/intelrealsense.py:275]: For consistency with OpenCVCamera, `self.camera_index` is now an int.

File: lerobot/common/robot_devices/cameras/intelrealsense.py (hunk 5 above)
Comment (Cadene) [reply lerobot/common/robot_devices/cameras/intelrealsense.py:150]: Updated `read` docstring

```
    def read(self, temporary_color: str | None = None) -> np.ndarray | tuple[np.ndarray, np.ndarray]:
        """Read a frame from the camera returned in the format height x width x channels (e.g. 480 x 640 x 3)
        of type `np.uint8`, contrarily to the pytorch format which is float channel first.

        When `use_depth=True`, returns a tuple `(color_image, depth_map)` with a depth map in the format
        height x width (e.g. 480 x 640) of type np.uint16.

```

**Review by Cadene — COMMENTED** [review]
Thanks for the reviews!

File: lerobot/common/robot_devices/cameras/intelrealsense.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Cadene) [reply lerobot/common/robot_devices/cameras/intelrealsense.py:112]: Changed `camera.camera_idx` to `camera.camera_index`

**Pull request merged**